File: samples/cext-trey-research-python/functions/consultants/consultants.py
import azure.functions as func
import logging
from services.consultant_api_service import consultant_api_service
from services.identity_service import identity_service
from services.utilities import clean_up_parameter
import json
import dataclasses

async def consultants(req: func.HttpRequest) -> func.HttpResponse:
    logging.info("HTTP trigger function consultants processed a request.")

    # Initialize the response JSON
    response_json = {
        "results": []
    }

    try:
        # Validate the request using IdentityService
        await identity_service.validate_request(req)
        # Get the input parameters from the request
        consultant_name = req.params.get('consultantName', '').lower()
        project_name = req.params.get('projectName', '').lower()
        skill = req.params.get('skill', '').lower()
        certification = req.params.get('certification', '').lower()
        role = req.params.get('role', '').lower()
        hours_available = req.params.get('hoursAvailable', '').lower()
        logging.info("Checking consultant_id from route parameters.")
        consultant_id = req.route_params.get('id')

        if consultant_id:
            consultant_id = consultant_id.lower()
            logging.info(f"➡️ GET /api/consultants/{consultant_id}: request for consultant {consultant_id}")
            result = await consultant_api_service.get_api_consultant_by_id(consultant_id)
            response_json['results'] = [vars(result)]
            logging.info(f"   ✅ GET /api/consultants/{consultant_id}: response status 1 consultant returned")
            return func.HttpResponse(
                body=json.dumps(response_json),
                mimetype="application/json",
                status_code=200
            )

        logging.info(f"➡️ GET /api/consultants: request for consultantName={consultant_name}, "
                     f"projectName={project_name}, skill={skill}, certification={certification}, "
                     f"role={role}, hoursAvailable={hours_available}")

        # Clean up the parameters
        consultant_name = clean_up_parameter("consultantName", consultant_name)
        project_name = clean_up_parameter("projectName", project_name)
        skill = clean_up_parameter("skill", skill)
        certification = clean_up_parameter("certification", certification)
        role = clean_up_parameter("role", role)
        hours_available = clean_up_parameter("hoursAvailable", hours_available)

        # Get consultants based on the filters
        result = await consultant_api_service.get_api_consultants(
            consultant_name, project_name, skill, certification, role, hours_available
        )
        response_json['results'] = [dataclasses.asdict(ApiConsultant) for ApiConsultant in result]
        logging.info(f"   ✅ GET /api/consultants: response status 200; {len(result)} consultants returned")
        return func.HttpResponse(
            body=json.dumps(response_json),
            mimetype="application/json",
            status_code=200
        )

    except Exception as e:
        logging.error(f"   ⛔ Returning error status code 500: {str(e)}")
        response_json['results'] = {
            "status": 500,
            "message": str(e)
        }
        return func.HttpResponse(
            body=json.dumps(response_json),
            mimetype="application/json",
            status_code=500
        )

File: samples/cext-trey-research-csharp/Functions/Consultants.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using Trey.Research.Declarative.Agent.Services;
using Trey.Research.Declarative.Agent.Utilities;
using Microsoft.Extensions.Configuration;

namespace Trey.Research.Declarative.Agent.Functions
{
    public class ConsultantsFunction
    {
        private readonly ConsultantApiService _consultantApiService;
        private readonly IdentityService _identityService;
        private readonly IConfiguration _configuration;

        public ConsultantsFunction(ConsultantApiService consultantApiService, IdentityService identityService, IConfiguration configuration)
        {
            _consultantApiService = consultantApiService;
            _identityService = identityService;
            _configuration = configuration;
        }

        [FunctionName("Consultants")]
        public async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "consultants/{id?}")] HttpRequest req,
            string id,
            ILogger log)
        {
            log.LogInformation("HTTP trigger function consultants processed a request.");

            var jsonResponse = new { results = new List<object>() };

            try
            {
                // Initialize identity
                var userInfo = await new IdentityService(_configuration).ValidateRequest(req);

                // Get the input parameters
                var consultantName = req.Query.ContainsKey("consultantName") ? req.Query["consultantName"].ToString().ToLower() : "";
                var projectName = req.Query.ContainsKey("projectName") ? req.Query["projectName"].ToString().ToLower() : "";
                var skill = req.Query.ContainsKey("skill") ? req.Query["skill"].ToString().ToLower() : "";
                var certification = req.Query.ContainsKey("certification") ? req.Query["certification"].ToString().ToLower() : "";
                var role = req.Query.ContainsKey("role") ? req.Query["role"].ToString().ToLower() : "";
                var hoursAvailable = req.Query.ContainsKey("hoursAvailable") ? req.Query["hoursAvailable"].ToString().ToLower() : "";

                if (!string.IsNullOrEmpty(id))
                {
                    log.LogInformation($"➡️ GET /api/consultants/{id}: request for consultant {id}");
                    var result = await _consultantApiService.GetApiConsultantById(id.ToLower());
                    jsonResponse = new { results = new List<object> { result } };
                    log.LogInformation($"   ✅ GET /api/consultants/{id}: response status 1 consultant returned");
                    return new OkObjectResult(jsonResponse);
                }

                log.LogInformation($"➡️ GET /api/consultants: request for consultantName={consultantName}, projectName={projectName}, skill={skill}, certification={certification}, role={role}, hoursAvailable={hoursAvailable}");

                // Clean up parameters
                consultantName = Utility.CleanUpParameter("consultantName", consultantName);
                projectName = Utility.CleanUpParameter("projectName", projectName);
                skill = Utility.CleanUpParameter("skill", skill);
                certification = Utility.CleanUpParameter("certification", certification);
                role = Utility.CleanUpParameter("role", role);
                hoursAvailable = Utility.CleanUpParameter("hoursAvailable", hoursAvailable);

                
                var results = await _consultantApiService.GetApiConsultants(consultantName, projectName, skill, certification, role, hoursAvailable);
                jsonResponse = new { results = results.Cast<object>().ToList() };
                log.LogInformation($"   ✅ GET /api/consultants: response status OK; {results.Count} consultants returned");

                return new OkObjectResult(jsonResponse);
            }
            catch (Exception ex)
            {
                log.LogError($"   ⛔ Returning error status code 500: {ex.Message}");
                return new ObjectResult(new { error = ex.Message }) { StatusCode = 500 };
            }
        }
    }
}